post: Setting up PCP and Grafana metrics with Cockpit (#429)
cockpit-project/cockpit-project.github.io: _posts/2021-07-05-pcp-grafana.md

---
title: Setting up PCP and Grafana metrics with Cockpit
author: pitti
date: 2021-07-05
category: howto
tags: metrics pcp grafana
---

> Debugging performance problems is a common troubleshooting activity. Monitoring the usage of CPU, memory, network, and other resources helps administrators to spot patterns when unusual resource usage happens.

Reviewer: ```suggestion
Finding performance problems is a common troubleshooting activity. Monitoring the usage of CPU, memory, network, and other resources helps administrators to spot patterns when unusual resource usage occurs.
```

> Cockpit can help you with this. Let's walk through!

Reviewer: ```suggestion
Cockpit can help. Let's take a look!
```

> Cockpit can help you with this. Let's walk through!

Reviewer: ```suggestion
Cockpit can help. This walk-through shows how.
```

## Starting point

> Let's start with a pristine Fedora 34 system which has at least [Cockpit 247](https://cockpit-project.org/blog/cockpit-247.html). Everything here works in principle on CentOS/RHEL 8, Debian, and Ubuntu as well, but at the time of this writing these distributions still have an older Cockpit version. Let's make sure that Cockpit is installed, enabled, and opened in the firewall, as [documented on the Install page](https://cockpit-project.org/running.html):

Reviewer: ```suggestion
Let's start with a pristine Fedora 34 system which has at least [Cockpit 247](https://cockpit-project.org/blog/cockpit-247.html). Everything here works in principle on CentOS/RHEL 8, Debian, and Ubuntu as well, but at the time of writing these distributions still have an older Cockpit version. Let's make sure that Cockpit is installed, enabled, and that the port is open in the firewall, as [documented on the Install page](https://cockpit-project.org/running.html):
```

```
# rpm -q cockpit
cockpit-247-1.fc34.x86_64
# systemctl enable --now cockpit.socket
# firewall-cmd --permanent --add-service cockpit
# firewall-cmd --reload
```

## Metrics in Cockpit for the local machine

> Then you can log into Cockpit on [https://localhost:9090](https://localhost:9090), or the corresponding host name of the system which you want to troubleshoot. The Overview page shows a very brief CPU/memory usage with a "View Details and history" link:

Reviewer: ```suggestion
You can now log into Cockpit on [https://localhost:9090](https://localhost:9090), or the corresponding host name of the system which you want to troubleshoot. The Overview page shows a very brief CPU/memory usage with a "View Details and history" link:
```

![Overview usage card](/images/overview-card-usage.png)

That link leads you to the Metrics page, which in a minimal installation only shows current metrics:

![Metrics page without PCP](/images/metrics-page-no-pcp.png)

> For spotting performance problems in the past, we need to enable historical metrics, which are provided by [Performance Co-Pilot](https://pcp.io/), aka.  "PCP".

Reviewer: ```suggestion
For spotting performance problems in the past, we need to enable historical metrics, which are provided by [Performance Co-Pilot](https://pcp.io/), aka. "PCP".
```

> The "Install cockpit-pcp" button installs the necessary support for this, and asks you to log in again (the new cockpit-pcp bridge cannot be used in an existing session). After that, the Metrics page will start to show historical metrics. Combined with Cockpit's other pages such as "Services", "Logs", or "Podman containers", this is a very powerful tool to find performance related problems:

Reviewer: ```suggestion
The "Install cockpit-pcp" button installs the necessary support for this, and asks you to log in again (the newly-installed cockpit-pcp support cannot be used in an existing session). After that, the Metrics page will start to show historical metrics. Combined with Cockpit's other pages such as "Services", "Logs", or "Podman containers", this is a very powerful tool to find performance related problems:
```

<iframe width="960" height="720" src="https://www.youtube.com/embed/8_NLoJBjay4?rel=0" frameborder="0" allowfullscreen></iframe>

## Metrics of multiple machines

But what if you want to monitor not just one machine, but your whole fleet?

> Cockpit used to have a "Dashboard" to see some basic resource usage of up to twenty machines, but this was very inflexible, inefficient, insecure, lacking important features such as alerting, and was bound to a running Cockpit session. That's why the Dashboard got removed in [Cockpit 234](https://cockpit-project.org/blog/cockpit-234.html).

Reviewer: ```suggestion
Cockpit used to have a "Dashboard" to see some basic resource usage of up to twenty machines, but this was very inflexible, inefficient, insecure, lacking important features such as alerting, and was bound to a running Cockpit session. For these reasons, the Dashboard was removed in [Cockpit 234](https://cockpit-project.org/blog/cockpit-234.html).
```

> One of today's industry standard solutions for visualizing metrics is [Grafana](https://grafana.com/); it does not have above limitations. PCP offers a [Grafana plugin](https://grafana-pcp.readthedocs.io/en/latest/) to provide the machine's metrics to a local or remote Grafana installation.

Reviewer: ```suggestion
One of today's industry standard solutions for visualizing metrics is [Grafana](https://grafana.com/); it doesn't have above limitations. PCP offers a [Grafana plugin](https://grafana-pcp.readthedocs.io/en/latest/) to provide the machine's metrics to a local or remote Grafana installation.
```

Author: Out of interest, I thought contractions were informal, and should be avoided in written text?

## Setup of monitored machines

Setting up the machines for sending metrics to Grafana involves a couple of steps:

 * Install PCP and enable `pmlogger.service` to start data collection. This already happened with the "Install cockpit-pcp" button above.
 * Enable `pmproxy.service`, the PCP component which exposes a read-only metrics query API to the network.
 * pmproxy needs the [Redis database](https://redis.io/), so the `redis` package needs to be installed, and the service configured to start with `pmproxy.service`.

>  * pmproxy opens a network port (44322 by default), which needs to be opened in the firewall for the zone that contains the Grafana machine.

Reviewer: While I'm reading this entire section, at each separate point, I'm asking myself "...uh.. but how do I do that?  Where's the CLI fragment like was provided above for installing cockpit?  Do I have to figure this out on my own?  Should I open a terminal now and poke around?  Try Google?"

...of course, the answer to all of these questions is just below, but it might save the user some frustration to put it up top and use this list to describe all of the things that will happen as a result of the magic button.

Author: Good point, thanks for bringing it up! I restructured this paragraph a bit to make this obvious from the start. WDYT?

Reviewer: That helps a lot, thanks :)

The most recent Cockpit version can do all these steps for you with the "Metrics settings" button at the top of the page. That opens a dialog where you can enable "Export to network":

![Metrics settings with enabled pmproxy](/images/metrics-settings-2.png)

> This will let you confirm any extra package that needs to be installed (such as redis). After that, you may get an alert for opening the corresponding firewall service:

Reviewer: ```suggestion
This will let you confirm any extra packages that need to be installed (such as redis). After that, you may get an alert for opening the corresponding firewall service:
```
